<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>RNAseq read alignment with STAR at OSC</title>
    <meta charset="utf-8" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/default-fonts.css" rel="stylesheet" />
    <link rel="stylesheet" href="slides.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">

class:inverse middle center



# RNAseq read alignment with STAR at OSC

----

&lt;br&gt; &lt;br&gt; &lt;br&gt;

### Jelmer Poelstra, MCIC Wooster
### 2021/02/17 (updated: 2021-02-16)
  
---

## Overview of the analysis pipeline

&lt;p align="center"&gt;
&lt;img src=img/pipeline-overview2.png width="90%"&gt;
&lt;/p&gt;

---

## Aligning RNAseq reads

The purpose of aligning reads, also known as *mapping*:  
**determine where in the genome each read originates from.**
  
Aligning RNAseq reads is more challenging than DNA reads:
due to *splicing*, many reads can't be simply be mapped wholly to the genome.
  
Moreover, there is often *alternative splicing*,
such that a single "correct" reference is simply not applicable.

&lt;figure&gt;
&lt;p align="center"&gt;
&lt;img src="img/DNA_alternative_splicing.gif" width="75%"&gt;
&lt;figcaption&gt;Figure from &lt;a href="https://en.wikipedia.org/wiki/Alternative_splicing#/media/File:DNA_alternative_splicing.gif"&gt;Wikipedia&lt;/a&gt;&lt;/figcaption&gt;
&lt;/p&gt;
&lt;/figure&gt;

---

## Aligning RNAseq reads (cont.)

Three different strategies are possible &amp;ndash; but mapping to the genome
is preferable as long as a reference genome is available:

&lt;br&gt;

&lt;figure&gt;
&lt;p align="center"&gt;
&lt;img src="img/mapping-strategies.png" width="85%"&gt;
&lt;figcaption&gt;Figures from &lt;a href="https://www.nature.com/articles/nmeth1010-793"&gt;Cloonan &amp; Grimmond 2010&lt;/a&gt;&lt;/figcaption&gt;
&lt;/p&gt;
&lt;/figure&gt;

---

## Aligning RNAseq reads (cont.)

Therefore, RNAseq mappers have to be "splice-aware":

&lt;br&gt;

&lt;figure&gt;
&lt;p align="center"&gt;
&lt;img src="img/star-map.png" width="80%"&gt;
&lt;figcaption&gt;Figure from Dobin et al. 2013&lt;/figcaption&gt;
&lt;/p&gt;
&lt;/figure&gt;

---

## Alignment output: SAM/BAM files

- SAM/BAM files contain the original reads *along with the genomic coordinates
  to which they were mapped* (and more information).

- We get 1 SAM/BAM file for each sample with both the forward and the reverse
  reads (not 2 separate files like for FASTQ files).

--

- SAM is an uncompressed text file and BAM is its binary, compressed counterpart.
  
  - SAM: Sequence Alignment/Map
  
  - BAM: Binary Alignment/Map

  - Most software can work directly with BAM files, and you can use programs
    like Samtools to view them.
  
- For subsequent analyses, BAM files usually also need to be *sorted* by
  genomic coordinates (e.g. with Samtools but also STAR).

---

## The STAR aligner

**About STAR:**

- STAR is an acronym for "Spliced Transcripts Alignment to a Reference".

- Reference paper &amp;ndash; Dobin et al. 2013, *Bioinformatics*: [STAR: ultrafast
  universal RNA-seq aligner](https://academic.oup.com/bioinformatics/article/29/1/15/272537).

- [This page](https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/03_alignment.html)
  has a nice quick, visual explanation of STAR's alignment strategy.

- It's a very fast aligner, but memory-intensive. Luckily we have OSC!

--

&lt;br&gt;

.content-box-info[
Before we can perform the mapping, we also need to index the genome with STAR.
Every mapper has its own mapping algorith and associated with that also its
own way to create a genome dictionary.
]

---

## STAR at OSC

- STAR is available in a module at OSC. The Owens cluster has a more recent
  version available, so make sure to log in to the Owens cluster.

  ```sh
  $ cd /fs/project/PAS0471/teach/misc/2021-02_rnaseq/$USER
  ```
  
  ```sh
  $ module load star/2.6.0a
  ```

---

## STAR options for genome indexing

Some options when generating a genome index with STAR are as follows:

| Option                        | Meaning
|-------------------------------|---------
| `--runThreadN 18`             | Use 18 cores/threads
| `--runMode genomeGenerate`    | The mode to index a genome
| `--genomeDir mydir`           | Destination dir for index 
| `--genomeFastaFiles my.fasta` | Path to the genome FASTA file
| `--sjdbGTFfile my.gtf`        | Path to the genome GTF file
| `--sjdbOverhang ReadLength-1` | Length of the donor/acceptor sequence &lt;br&gt; on each side of the junctions
| `--genomeSAindexNbases 14`    | Length of the SA pre-indexing string

--

.content-box-info[
According to the documentation `genomeSAindexNbases` should be:

`log2(GenomeLength)/2 - 1)` =&gt; `log2(782520133)/2 - 1)` = 13.77 = 14
]

---

## The key parts of a script to index the genome

```sh
#!/bin/bash
#SBATCH --account=PAS0471
#SBATCH --time=2:00:00
#SBATCH --cpus-per-task=18
module load star/2.6.0a

STAR --runThreadN "$SLURM_CPUS_ON_NODE" \
    --runMode genomeGenerate \
    --genomeDir "$index_dir" \
    --genomeFastaFiles "$ref_fa" \
    --sjdbGTFtagExonParentTranscript "$ref_gff" \
    --genomeSAindexNbases 13 \
    --sjdbOverhang ReadLength-1
```

--

Submit the script:

```sh
$ ref_fa=data/ref/Heinz1706_4.00/S_lycopersicum_chromosomes.4.00.fa
$ ref_gff=data/ref/annot/ITAG4.1/ITAG4.1_gene_models.gff
$ index_dir=data/ref/Heinz1706_4.00/STAR_index

$ sbatch scripts/align/star_index.sh \
    "$ref_fa" "$ref_gff" "$index_dir"
#&gt; Submitted batch job 12826056
```

---

## Check the output

```sh
$ ls
#&gt; slurm-STAR-index-12826056.out Log.out

$ cat slurm-STAR-index-12826056.out
#&gt; Feb 10 14:53:19 ..... started STAR run
#&gt; Feb 10 14:53:19 ... starting to generate Genome files
#&gt; Feb 10 14:53:34 ... starting to sort Suffix Array. This may take a long time...
#&gt; Feb 10 14:53:39 ... sorting Suffix Array chunks and saving them to disk...
#&gt; Feb 10 14:55:09 ... loading chunks from disk, packing SA...
#&gt; Feb 10 14:55:28 ... finished generating suffix array
#&gt; Feb 10 14:55:28 ... generating Suffix Array index
#&gt; Feb 10 14:56:15 ... completed Suffix Array index
#&gt; Feb 10 14:56:15 ... writing Genome to disk ...
#&gt; Feb 10 14:56:15 ... writing Suffix Array to disk ...
#&gt; Feb 10 14:56:18 ... writing SAindex to disk
#&gt; Feb 10 14:56:18 ..... finished successfully
```

---

## Check the output (cont.)

STAR also automatically outputs a file `Log.out` with lots of information
about the run:
  
```sh
$ less Log.out

#&gt; STAR version=STAR_2.6.0a
#&gt; STAR compilation time,server,dir=Mon Apr 23 13:19:26 EDT 2018 #&gt; florence.cshl.edu:/sonas-hs/gingeras/nlsas_norepl/user/dobin/STAR/STAR.master/source
#&gt; ##### DEFAULT parameters:
#&gt; versionSTAR                       20201
#&gt; versionGenome                     20101   20200   
#&gt; parametersFiles                   -   
#&gt; sysShell                          -
#&gt; runMode                           alignReads
#&gt; runThreadN                        1
#&gt; runDirPerm                        User_RWX
#&gt; runRNGseed                        777
#&gt; ...
```

```sh
$ wc -l Log.out
#&gt; 416
```

---

## Check the output (cont.)

- Check the actual indexing output:
  
  ```sh
  $ ls -lh data/ref/Heinz1706_4.00/STAR_index
  
  #&gt; total 7.2G
  #&gt; -rw-r--r-- 1 jelmer PAS0471  116 Feb 10 14:53 chrLength.txt
  #&gt; -rw-r--r-- 1 jelmer PAS0471  246 Feb 10 14:53 chrNameLength.txt
  #&gt; -rw-r--r-- 1 jelmer PAS0471  130 Feb 10 14:53 chrName.txt
  #&gt; -rw-r--r-- 1 jelmer PAS0471  130 Feb 10 14:53 chrStart.txt
  #&gt; -rw-r--r-- 1 jelmer PAS0471 748M Feb 10 14:56 Genome
  #&gt; -rw-r--r-- 1 jelmer PAS0471  734 Feb 10 14:56 genomeParameters.txt
  #&gt; -rw-r--r-- 1 jelmer PAS0471 6.1G Feb 10 14:56 SA
  #&gt; -rw-r--r-- 1 jelmer PAS0471 374M Feb 10 14:56 SAindex
  ```

---

## STAR options for alignment

Some options for aligning reads to the genome using STAR are  
(_not showing options also shown for the indexing_):

| Option                    | Meaning
|---------------------------|------------
| `--readFilesIn`           | Path to FASTQ file
| `--outFileNamePrefix`     | Path and (**sample**) prefix for output files
| `--outFilterMultimapNmax` | Max. nr. alignments for a read &lt;br&gt; (default: 10)
| `--outSAMtype` | Output sorted BAM (default: SAM) &lt;br&gt; `BAM SortedByCoordinate` for sorted BAM
| `--outSAMunmapped`        | What to do with unmapped reads &lt;br&gt; (default: don't output)

--

.content-box-warning[
Check maximum intron size in GFF:

`--alignIntronMin 20` &amp;mdash; minimum intron length

`--alignIntronMax 1000000` &amp;mdash; maximum intron length
]

---

## The key parts of a script to map reads for 1 sample

```sh
#!/bin/bash
#SBATCH --account=PAS0471
#SBATCH --time=2:00:00
#SBATCH --cpus-per-task=18
module load star/2.6.0a

R1=$(ls "$fastq_dir"/*"$sample"*R1*fastq.gz)
R2=$(ls "$fastq_dir"/*"$sample"*R2*fastq.gz)

STAR --runThreadN "$SLURM_CPUS_ON_NODE" \
   --genomeDir "$index_dir" \
   --sjdbGTFtagExonParentTranscript "$ref_gff" \
   --readFilesIn "$R1" "$R2" \
   --readFilesCommand zcat \
   --outFileNamePrefix "$bam_dir/$sample"_ \
   --outSAMtype BAM SortedByCoordinate
```

--

Submit the script:

```sh
$ sample=C6_myb2
$ fastq_dir=data/fastq_concat
$ bam_dir=results/align/bam
$ sbatch -o "$slurm_log" scripts/align/star_align.sh \
      "$sample" "$fastq_dir" "$index_dir" "$ref_gff" "$bam_dir"
```

---

## Check the output

```sh
$ cat slurm-align-C6_myb2-12836540.out
#&gt; Starting script star_align.sh
#&gt; Thu Feb 11 20:24:21 EST 2021
#&gt; -------------------

#&gt; R1 input file: data/fastq_concat/C6_myb2_R1.fastq.gz
#&gt; R2 input file: data/fastq_concat/C6_myb2_R2.fastq.gz
#&gt; Genome index dir: data/ref/Heinz1706_4.00/STAR_index
#&gt; Genome annotation file: data/ref/annot/ITAG4.1/ITAG4.1_gene_models.gff
#&gt; Bam (output) directory: results/align/bam

#&gt; Running STAR....
#&gt; Feb 11 20:24:21 ..... started STAR run
#&gt; Feb 11 20:24:21 ..... loading genome
#&gt; Feb 11 20:24:27 ..... started mapping
```

---

## Check the output (cont.)

```sh
$ ls $bam_dir
```

---

## A script to loop over all files

```sh
#!/bin/bash

samples=($(cat "$sample_list"))

for sample in "${samples[@]}"; do

  R1=$(ls "$fastq_dir"/*"$sample"*_R1*fastq.gz)

  slurm_log=slurm-align-"$sample"-%j.out 

  echo "Submitting star_align.sh for $sample"
  sbatch -o "$slurm_log" scripts/align/star_align.sh \
      "$sample" "$fastq_dir" "$index_dir" "$ref_gff" "$bam_dir"

done
```

---

## Submit the loop script

```sh
$ ./scripts/align/star_align_loop.sh
    "$sample_list" "$fastq_dir" "$index_dir" "$ref_gff" "$bam_dir"
    
#&gt; Starting script star_align_loop.sh
#&gt; Wed Feb 10 21:03:13 EST 2021
#&gt; -------------------
#&gt; 
#&gt; FASTQ dir: data/fastq_concat
#&gt; Sample list: data/meta/samples.txt
#&gt; Genome index dir: data/ref/Heinz1706_4.00/STAR_index
#&gt; Genome annotation file: data/ref/annot/ITAG4.1/ITAG4.1_gene_models.gff
#&gt; Bam (output) directory: results/align/bam
#&gt; 
#&gt; Submitting star_align.sh for C6_myb2
#&gt; Submitted batch job 12830287
#&gt; Submitting star_align.sh for C6_myb2
#&gt; Submitted batch job 12830289
#&gt; Submitting star_align.sh for C6_myb3
#&gt; Submitted batch job 12830291
#&gt; Submitting star_align.sh for C6_myb3
#&gt; Submitted batch job 12830293
#&gt; ...
```

---

## Checking the output

```sh
$ grep "Uniquely mapped" results/align/bam/star_logs/*Log.final*
#&gt; C6_myb2_Log.final.out: Uniquely mapped reads % |   76.30%
#&gt; C6_myb3_Log.final.out: Uniquely mapped reads % |   79.46%
#&gt; C6_myb4_Log.final.out: Uniquely mapped reads % |   56.44%
#&gt; cnt_1_Log.final.out:   Uniquely mapped reads % |   71.11%
#&gt; cnt_2_Log.final.out:   Uniquely mapped reads % |   78.40%
#&gt; cnt_3_Log.final.out:   Uniquely mapped reads % |   66.60%
#&gt; CT_05_Log.final.out:   Uniquely mapped reads % |   80.29%
#&gt; CT_06_Log.final.out:   Uniquely mapped reads % |   81.05%
#&gt; CT_09_Log.final.out:   Uniquely mapped reads % |   86.46%
#&gt; myb_1_Log.final.out:   Uniquely mapped reads % |   87.14%
#&gt; myb_2_Log.final.out:   Uniquely mapped reads % |   86.76%
#&gt; myb_3_Log.final.out:   Uniquely mapped reads % |   87.18%
#&gt; T_26_Log.final.out:    Uniquely mapped reads % |   85.01%
#&gt; T_29_Log.final.out:    Uniquely mapped reads % |   q86.01%
```

---

## Running MultiQC on the STAR logs

```sh
$ sbatch scripts/qc/multiqc.sh results/align/bam/star_logs
```
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "rainbow",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
