---
class: inverse middle center
name: fastqc-output

# FastQC output: "module" by module

----

<br>

.left[
### Two useful sources of information: 
- ### [FastQC page for each module](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/) (Looks weird but works!)
- ### [MSU FastQC tutorial and FAQ](https://rtsf.natsci.msu.edu/sites/_rtsf/assets/File/FastQC_TutorialAndFAQ_080717.pdf)
]

<br> <br>

---

## Summary and overview of modules

<p align="left">
<img src=img/fastqc_summary.png width="35%">
</p>

---

## Module 1: Basic statistics 

<p align="center">
<img src=img/fastqc_mod01_basic-stats.png width="100%">
</p>

---

## Module 2: Per-base quality along the read 

.pull-left[
**Good / OK:**

<p align="center">
<img src=img/fastqc_good.png width="100%">
</p>
]

.pull-right[
**Bad:**

<p align="center">
<img src=img/fastqc_bad.png width="100%">
</p>
]

.content-box-info[
- A decrease in sequence quality along the reads is normal.
- R2 (reverse) reads are usually worse than R1 (forward) reads.
]

---

## Module 3: Per-tile sequence quality

.pull-left[
**Good:**

<p align="center">
<img src=img/fastqc_mod03_per-tile_good.png width="100%">
</p>
]

.pull-right[
**Bad:**

<p align="center">
<img src=img/fastqc_mod03_per-tile_bad.png width="100%">
</p>
]



---

## Module 4: Per-sequence quality scores

.pull-left[
**Good:**

<p align="center">
<img src=img/fastqc_mod04_per-seq-qual_good.png width="100%">
</p>
]

.pull-right[
**Bad:**

<p align="center">
<img src=img/fastqc_mod04_per-seq-qual_bad.png width="100%">
</p>
]


---

## Module 5: Per-base sequence content

.pull-left[
**Good:**

<p align="center">
<img src=img/fastqc_mod05_per-base-content_good.png width="90%">
</p>
]

.pull-right[
**Bad:**

<p align="center">
<img src=img/fastqc_mod05_per-base-content_bad.png width="90%">
</p>
]

> It's worth noting that some types of library will always produce biased sequence composition, normally at the start of the read. Libraries produced by priming using random hexamers (including nearly all RNA-Seq libraries) and those which were fragmented using transposases inherit an intrinsic bias in the positions at which reads start. Whilst this is a true technical bias, it isn't something which can be corrected by trimming and in most cases doesn't seem to adversely affect the downstream analysis. It will however produce a warning or error in this module.
> &mdash; [source](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/3%20Analysis%20Modules/4%20Per%20Base%20Sequence%20Content.html)

---

## Module 6: Per-sequence GC content

.pull-left[
**Good:**

<p align="center">
<img src=img/fastqc_mod06_GC_good.png width="100%">
</p>
]

.pull-right[
**Bad:**

<p align="center">
<img src=img/fastqc_mod06_GC_bad.png width="100%">
</p>
]

<br>

.content-box-info[
The expected distribution is for whole-genome shotgun sequencing &ndash;
it is normal for RNA-seq data to have a narrower distribution.
]

---

## Module 7: Per-base N content

.pull-left[
**Good:**

<p align="center">
<img src=img/fastqc_mod07_N_good.png width="100%">
</p>
]

.pull-right[
**Bad:**

<p align="center">
<img src=img/fastqc_mod07_N_bad.png width="100%">
</p>
]

---

## Module 8: Sequence length distribution

- Will throw a warning as soon as not all sequences are of the same length,
  but this is quite normal.
  
---

## Module 9: Sequence duplication levels

.pull-left[
**Good:**

<figure>
<p align="center">
<img src=img/fastqc_mod09_seqdup_good.png width="100%">
<figcaption><a href="https://rtsf.natsci.msu.edu/sites/_rtsf/assets/File/FastQC_TutorialAndFAQ_080717.pdf">Figure source</a> </figcaption>
</p>
</figure>
]

.pull-right[
**"Bad":**

<p align="center">
<img src=img/fastqc_mod09_seqdup_bad.png width="100%">
</p>

]

<br>

- Will usually throw a warning for RNA-seq data, which can be ignored:
  highly expressed transcripts will be present in the dataset many times.
  
- Pay attention to the blue line &ndash; the red line can generally be ignored.

---

## Module 10: Overrepresented sequences

<p align="center">
<img src=img/fastqc_mod10_overrepresented.png width="100%">
</p>

- Returns a *Warning* if any sequence is >0.1% of total.
- Returns *Failure* if any sequence is >1% of total.

---

## Module 11: Adapter content

.pull-left[
**Good:**

<p align="center">
<img src=img/fastqc_mod11_adapter_good.png width="100%">
</p>
]

.pull-right[
**Bad:**

<figure>
<p align="center">
<img src=img/fastqc_mod11_adapter_bad.png width="100%">
<figcaption><a href="https://rtsf.natsci.msu.edu/sites/_rtsf/assets/File/FastQC_TutorialAndFAQ_080717.pdf">Figure source</a> </figcaption>
</p>
</figure>
]

- When some of the insert sizes are shorter than the read length,
  adapters can end up in the sequence and should be remove.
